### ISSUE TITLE
No obvious synchronization mechanism

### ISSUE THREAD

**Issue by haarp** [stefan-niedermann/OwnCloud-Notes#64]
I'm loving this nice and minimalistic app!

But it lacks something: Synchronization. Currently, the app seems to synchronize with the server in irregular intervals. If I edit a note on the server with another app (e.g. OwnCloud's web interface), the notes app does not notice. Closing and reopening the app a couple of times seems to eventually make it synchronize and show the new version.

Worse yet, if I edit a note that has been edited elsewhere, the app overwrites any changes and crashes afterwards, both of which obviously shouldn't happen.

My proposal:
- Investigate the crash bug ;)
- Show a refresh button in the app's main view that animates when a sync is in progress
- Sync automatically whenever the app or a note is opened

Thanks!

**Comment by stefan-niedermann** [comment]
Hi,

thank you for feedback.
- I will try to reproduce and fix the crash bug.
- This app uses a swipe-to-refresh-layout: try to swipe from top to bottom in the list view
- It should sync on open the list view, but i will do some research (maybe it works currently not, if the app has not been closed and runs in the background?)

**Comment by haarp** [comment]
Ah, indeed, swipe-to-refresh is possible. But the animation should play even if the refresh was triggered automatically, for two reasons:
- To let the user know that there is a swipe-to-refresh gesture
- To let the user know when the app has synced

Pressing the back button doesn't close the app, does it? That's what I've been using. Although it would probably be better if a refresh is triggered whenever the app gains focus, whether it just started or was brought up from the background

**Comment by stefan-niedermann** [comment]
>  But the animation should play even if the refresh was triggered automatically, for two reasons:
> - To let the user know that there is a swipe-to-refresh gesture
> - To let the user know when the app has synced

@jancborchardt you opinion to this point pls

**Comment by stefan-niedermann** [comment]
> Worse yet, if I edit a note that has been edited elsewhere, the app overwrites any changes and crashes afterwards, both of which obviously shouldn't happen.

@haarp could you please retest this with v0.5.2? Should be fixed.

@HeaDBanGer84 is currently refactoring the sync-mechanism in sync-service-branch. So i guess this will fix your problems once it has been finished.

The last open task are the animations on automatically triggered refreshs, where i want to wait for @jancborchardt opinion

**Comment by haarp** [comment]
@stefan-niedermann 
Yes, the crash appears to be fixed, thanks :) However, it still overwrites changes done via other sources.

**Comment by 9662** [comment]
FWIW, I agree that the current system is not particularly discoverable. [MyOwnNotes](/aykit/MyOwnNotes) uses a refresh-type icon on the title bar (whatever it's called in Android), like this:

Not updating:
![Not updating](https://cloud.githubusercontent.com/assets/13080040/12747079/3b2a5cc2-c9a2-11e5-972d-4b9c470dd904.png)

Updating:
![Updating](https://cloud.githubusercontent.com/assets/13080040/12747087/4ef55022-c9a2-11e5-932c-7e6e54e7e9e8.png)

It triggers an update when the application comes to the foreground, when connectivity is regained, when one deletes / adds a note, and at some regular interval.

I have been using that application for about a year now, including roaming and with slow connections (bad coverage, e.g., in the mountains) and never had a problem figuring out what it was doing, so I can vouch for that approach.

**Comment by jancborchardt** [comment]
Normally a sync should just happen. Most apps these days have pull-to-refresh, and an extra refresh button is not needed, not well integrated and also distracts from other functionality like search.

Other apps also do not show any refresh animation when something is being synced. The simple reason being that it’s not important. You only need to be confronted with sync stuff when it’s _not_ working. Not when it’s working perfectly fine, as people expect.

When looking for examples, please take a peek at Google Keep, Evernote, Simplenote and the like for well designed popular apps.

**Comment by haarp** [comment]
@jancborchardt 

Syncing with Notes can take quite a few seconds. When I open the app, I want to know if what I'm seeing is up-to-date or if the app is still working on it.

There are other reasons for having this animation:
- Make the fact that there's pull-to-refresh discoverable (other apps scroll a lot, this an user might accidently discover it. Unless you have a ton of notes, there is no scrolling in this app)
- Important functions should be evident
- Pull-to-refresh is nice, but it's not very common yet. Few users would think to simply try it

**Comment by 9662** [comment]
@jancborchardt 

> Most apps these days have pull-to-refresh

Yes, and that's an argument ad populum.

Pull to refresh is terrible from a usability point of view:
 * It is not discoverable, unless …
 * … you trigger that bastard by accident, e.g., when you're trying to scroll back to the top of a list, which …
 * … potentially costs money (e.g., roaming, satellite modems)

Very gimmicky, very fashionable, but utter rubbish. Especially when the space for a couple icons is already there. Not to mention using contextual menus.

> You only need to be confronted with sync stuff when it’s not working.

No mate. Syncing is an operation which has a cost, which can be significant (in terms of money and/or time) depending who and where you are. Not everyone lives in the first world, let us not design for the middle-class white male with a wife, two kids, a dog, and a mortgage, shall we? Let the user know, however discreetly, that something is happening.

**Comment by jancborchardt** [comment]
@9662 how does showing feedback about syncing relate to the app being useful for people outside the »first world«? There’s still no way to cancel or rate-limit the syncing. Which honestly would be quite overkill for a text-only app.
Also, the app syncs automatically anyway – rightfully so, because people expect their data to be up to date.

It really seems like this app is not the right one for your use-case. I suggest using the normal ownCloud app in combination with a text editor of your choice.
As you also said in https://github.com/stefan-niedermann/OwnCloud-Notes/issues/76#issuecomment-181813161 maybe a one-size-fits-all approach doesn’t work for some. But it surely does work for most.

And any customization is very work- and time-expensive on our end. So considering ownCloud is an open source project, you can still just modify it and run your own copy if you require such modifications. But some things will just not land in the main version because it’s too much work to maintain.

**Comment by 9662** [comment]
@jancborchardt 

> how does showing feedback about syncing relate to the app being useful for people outside the »first world«?

It lets you know that something of interest is happening, as pointed out by me and @haarp above.

> There’s still no way to cancel or rate-limit the syncing

The user can temporarily turn off the data connection or Wifi, as the case may be. At least on my Android phone there is also a “sync” feature that lets you tell applications whether they should hit the network or not. I do not know if this application complies with that or not—it's a separate issue in any case.

> It really seems like this app is not the right one for your use-case.

Beg pardon? What use case are you referring to?

> [….] some things will just not land in the main version because it’s too much work to maintain.

How is using an icon instead of a gesture interaction “too much work to maintain”? Yes, there is an additional up-front cost now that the requirement (to have user-triggered updates) has been implemented via gesture, but that is just a one-off cost related to having chosen initially what turned out to be a suboptimal alternative, as evidenced by user feedback and requirements analysis.

**Comment by stefan-niedermann** [comment]
Okay guys,

i am following this ticket since over a week now. Both sides presented their arguments, and both sides have *valid* arguments.

Though i have to make a decision. Sorry, but
- there will be **no** button to refresh like in MyOwnNotes
- the automatically triggered synchronization will not be visible

**Comment by haarp** [comment]
For what it's worth, just now I opened this app (after a few days of not using it) and saw missing notes. Which made me think the server was broken (as I had this problem before), so I booted my PC to double-check. Which revealed that the notes were there after all. Turns out that the app hasn't refreshed the notes in days, possibly weeks.

Functioning auto-refresh on start/gaining focus and visible refresh would have prevented this :(
